site/source/docs/api_reference/emscripten.h.rst

```diff
@@ -1410,6 +1414,60 @@ Functions
 
   .. note:: This requires building with ``-s ASYNCIFY_LAZY_LOAD_CODE``.
 
+.. c:function:: emscripten_fiber emscripten_fiber_create(em_arg_callback_func func, void *arg, void *stack, int stack_size)
+
+    Creates a new fiber context. Fibers are light cooperative threads of
+    execution that can be used to implement stackful coroutines.
+
+    `func` is the fiber's entry point. It will be called with the argument `arg`
+    in the fiber's context when it's first entered.
+
+    `stack` must point to a pre-allocated memory region that will be used as the
+    fiber's stack. `stack_size` is the size of that region in bytes. `stack`
+    should be aligned to at least 16 bytes.
+
+ .. note:: If `func` returns, the entire program will end, as if `main`
+    had returned. To avoid this, you can use `emscripten_fiber_swap` to jump to
+    another fiber.
+
+.. c:function:: emscripten_fiber emscripten_fiber_create_from_current_context()
+
+    Creates a new fiber that represents the current execution context. This is
+    needed in order to switch back from a fiber into the main context.
+
+ .. warning:: The fiber returned by this function is incomplete and can not be
+    switched into immediately. You must first switch to another fiber, passing
+    this one as the first argument to `emscripten_fiber_swap`. This will
+    complete the continuation, allowing you to switch back to it later.
+
+.. c:function:: void emscripten_fiber_free(emscripten_fiber fiber)
+
+    Frees memory and any JS-side bookkeeping associated with `fiber`. This does
+    not free the stack, that is your responsibility. It also does not "stop" or
+    corrupt a "running" fiber, as the fiber struct essentially only represents
+    a continuation.
+
+.. c:function:: void emscripten_fiber_swap(emscripten_fiber old_fiber, emscripten_fiber new_fiber)
+
+    Switches the execution context to the one represented by `new_fiber`.
+
+    `old_fiber` is updated with the current context just before the switch, such
+    as that switching back into it via another `emscripten_fiber_swap` would
+    appear to return from the original `emscripten_fiber_swap`.
+
+    If `old_fiber` == `new_fiber` or any of them is `NULL`, the behavior is
+    undefined.
```

Comment:
Why is `old_fiber` needed here?  Doesn't the runtime track the current running fiber somewhere?  Seem like ti would be simple to just take the target here?     It also avoid the problem of passing the wrong thing as old_fiber?  Prematurely old_fiber has to be the current running one anyway?

Reply:
Windows seems to just a target fibre here: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-switchtofiber

Reply:
The runtime does not track the currently running fiber. It does not really have a concept of a "running fiber" at all. In this implementation a fiber only represents a continuation, i.e. a bunch of state that describes how to set up the stack pointers and where to rewind the call stack in order to perform a context switch. Once that is done the runtime couldn't care less whether that state came from a fiber or not; you can even free a "running" fiber and nothing bad will happen (as long as you keep the stack alive).

Windows fibers are a little more high level than that. My fibers weren't modeled after them, and are more similar to POSIX [ucontext](https://linux.die.net/man/3/swapcontext). It's an explicit design decision to keep state to the bare minimum. This API is not designed to be used directly, but rather wrapped into a higher-level construct suitable to the application (asymmetrical coroutines being one example). I think it makes sense to shift the responsibility of tracking the current fiber to the application, because in virtually any conceivable scenario it will probably have to do that anyway.

Reply:
But why not track the current fibre?  Its trivial to do isn't it?  And then you can just remove `old_fibre` here?  Are there any contexts there is makes sense to swap with `old_fiber` not being the current one?   If not, I don't see why we wouldn't just track current fibre as a global.

Reply:
There's the question of how that's going to work with threads, but that's a can of worms I would rather keep tightly sealed for now. I'm not even sure how Asyncify works with threads.

But other than that I just view it as unnecessary bloat. The mechanism can function without tracking, so there's no tracking. I believe that any sort of non-trivial usage will require the application to track the necessary information anyway, even if the runtime does it on its own. That's why I find tracking in runtime redundant.

In other words, I prioritized simplicity of the implementation over simplicity of the interface here. I'm somewhat reluctant to change that, but I can do so if you insist.